Implement Wasm js-string builtins

ast-hugger opened the pull request:
#### 5c77b1f580e9552b8e1fac29879759ba0cd42a31
<pre>
Implement Wasm js-string builtins
<a href="https://bugs.webkit.org/show_bug.cgi?id=292630">https://bugs.webkit.org/show_bug.cgi?id=292630</a>
<a href="https://rdar.apple.com/150798623">rdar://150798623</a>

Reviewed by Keith Miller.

This patch implements the Wasm js-string builtin proposal
(<a href="https://github.com/WebAssembly/js-string-builtins).">https://github.com/WebAssembly/js-string-builtins).</a>

Availability of this feature is controlled by the `useWasmJSStringBuiltins` option,
`true` by default.

The relevant WPT test expectation files are changed to expect the tests to pass.

`FixedBitVector::test()` function is marked as `const` by this patch. It does
not require a mutable `this`, and the patch needs to use it on a vector
coming from a `const ModuleInformation&amp;`.

Canonical link: <a href="https://commits.webkit.org/298653@main">https://commits.webkit.org/298653@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de0cc00374e38fea839bfd0eb6fb6a438c9f219f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35796 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26341 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122191 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66691 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ffcc889f-8eca-4749-8279-5193977198da) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44384 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88243 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8cb797ef-1814-4e96-89f8-0f7ae3a249b7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119083 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29113 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104214 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68654 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28231 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22323 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65873 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108245 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98504 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22458 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125341 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/114663 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43029 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32311 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96974 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43394 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100403 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96758 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24628 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42028 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19917 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38976 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42916 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48508 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143360 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42383 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36955 "Found 8 new JSC stress test failures: wasm.yaml/wasm/stress/wasm-js-string-builtins.js.default-wasm, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-bbq, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-bbq-no-consts, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-collect-continuously, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-eager, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-eager-jettison, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-no-cjit, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-slow-memory (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45718 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44087 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -787,6 +787,37 @@ op(ipint_trampoline, macro ()
+    # For now this only supports the string builting proposal, and per the proposal
+    # all exceptions thrown are runtime errors, caused by an illegal argument.
+    throwException(IllegalArgument)
+end)
```
If we were going to do the per-builtin trampoline we talked about I think this would just be: 

```
macro buildWasmHostCallTrampoline(hostFunction, wasmInstanceArgGPR)
    functionPrologue()
    # This probably needs to be a store since you don't know how many arguments there are.
    store wasmInstance, wasmInstanceArgGPR
    storeMyGlobalCalleeForHostBuiltinsTODO()
    
    call hostFunction
    btpnz r1, .handleException

    functionEpilogue()
    ret

.handleExecption
    loadp JSWebAssemblyInstance::m_vm[wasmInstance], ws0
    storep cfr, VM::topCallFrame[ws0]
    # For now this only supports the string builting proposal, and per the proposal
    # all exceptions thrown are runtime errors, caused by an illegal argument.
    throwException(IllegalArgument)
end

global wasm_js_string_builtin_concat_trampoline
_wasm_js_string_builtin_concat_trampoline:
    buildWasmHostCallTrampoline(_wasmJsStringConcat, a2) # externref, externref, instance

global wasm_js_string_builtin_length_trampoline
_wasm_js_string_builtin_length_trampoline:
    buildWasmHostCallTrampoline(_wasmJsStringConcat, a1) # externref, instance

```

kmiller68 replied:
I put the instance at the end since these will all have the first n arguments filled by the wasm calling convention, so shuffling it to the front seems not ideal.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmModuleInformation.cpp:
```diff
@@ -39,6 +39,28 @@ ModuleInformation::ModuleInformation()
+    if (m_importedStringConstants)
+        ASSERT(m_importedStringConstants.value().impl()->refCount() == 1);
+    for (auto& string : m_qualifiedBuiltinSetNames) {
+        UNUSED_VARIABLE(string); // may be unused in assert
```
nit: assert_unused

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,825 @@
+
+    // The fact that 'array' is actually an (array mut i16) is guaranteed by Wasm type rules,
+    // or by a runtime check in JS entrypoint wrapper.
+    auto* array = static_cast<JSWebAssemblyArray*>(arrayObject);
```
Should be a `jsCast<JSWebAssemblyArray*>(arrayObject)` which will assert it's a JSWebAssemblyArray. We should also probably have an `ASSERT(isMutI16Array(array))` for good measure too.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,825 @@
+    JSString* string = stringArg.toStringOrNull(globalObject);
+    // The fact that 'array' is actually an (array mut i16) is guaranteed by Wasm type rules,
+    // or by a runtime check in JS entrypoint wrapper.
+    auto* array = static_cast<JSWebAssemblyArray*>(arrayObject);
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,825 @@
+    if (!arrayObject || !stringArg.isString()) [[unlikely]]
+        THROW_EXCEPTION;
+
+    JSString* string = stringArg.toStringOrNull(globalObject);
```
This can just be `jsCast<JSString*>(stringArg)`.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,825 @@
+
+#define IF_EXCEPTION_RETHROW_AND_RETURN(scope) \
+    if (scope.exception()) \
+        THROW_EXCEPTION;
```
This doesn't seem right for the JS API. We should just rethrow the error with `RETURN_IF_EXCEPTION(scope, { });`

ast-hugger replied:
I renamed the macro to a more descriptive `IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT`. This is the typical exception regrowing pattern, when a piece of logic catches an exception and rethrows it as a different one to expose a consistent exception surface to its callers. In our case we normally throw an exception saying "Illegal argument" when an argument is bad, and it should be the same no matter how we discover that the argument is bad.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,825 @@
+        double doubleVar = valueVar.toIntegerOrInfinity(globalObject); \
+        IF_EXCEPTION_RETHROW_AND_RETURN(scope); \
+        i32Var = static_cast<int32_t>(doubleVar); \
+    }
```
I don't think you need this "call"s to this should just be:

```
i32Var  = valueVar.toInt32(globalObject);
RETURN_IF_EXCEPTION(scope, { });
```

I think this also risks UB since casting a double to int32 is undefined if the truncated value cannot be represented in the destination type.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,825 @@
+        auto* tmpArray = static_cast<JSWebAssemblyArray*>((var).getObject()); \
+        if (!tmpArray || !isMutI16Array(tmpArray)) \
+            THROW_EXCEPTION; \
+    }
```
This seems very wrong since the object could be any object not a `JSWebAssemblyArray*`. So the downcast will covert a potentially attacker controlled JSObject into a `JSWebAssemblyArray` then load some bytes from its structure for the `isMutI16Array` check.

This should be something like:

```
auto* tmpArray = jsDynamicCast<JSWebAssemblyArray*>(jsValue);
if (!tmpArray || !isMutI16Array(tmpArray))
    THROW_EXCEPTION;    
```

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,825 @@
+
+// externref (externref)
+
+#define DEFINE_BUILTIN_ENTRY_R_R(setName, builtinName) \
```
We should at least have a comment defining these one letter values. But ideally we'd just write out something that doesn't require a lookup. e.g. `I` -> `I32`, `A` -> `MutI16Array`. In particular, I had to lookup what `R` was in the spec (I guess `Externref`). So e.g. `DEFINE_BUILTIN_ENTRY_R_RR` would be something like `DEFINE_BUILTIN_ENTRY_Externref__Externref_Externref`

ast-hugger replied:
I borrowed this scheme from `wasm-module-builder.js`, which is used by WPT tests (I think it comes from Google). I generally do like writing things out, but I'm afraid in this case it will make the identifiers so long it will be hard to tell them apart. I added a comment defining the type codes.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -40,6 +40,8 @@
 #include "WasmCallingConvention.h"
 #include "WasmModuleInformation.h"
 
+#include "js/WebAssemblyBuiltin.h"
```
You shouldn't need the "js/" prefix.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
```diff
@@ -118,7 +119,36 @@ Synchronousness WebAssemblyModuleRecord::link(JSGlobalObject* globalObject, JSVa
+    info->boxedWasmCalleeLoadLocation = &info->boxedCallee;
+    info->importFunctionStub = CodePtr<CFunctionPtrTag>::fromTaggedPtr((void*) builtin->implementation()).retagged<WasmEntryPtrTag>();
+    info->entrypointLoadLocation = &info->importFunctionStub;
+    info->targetInstance.set(vm, instance.get(), instance.get()); // FIXME(vb) what is the right owner to use here?
```
Yeah, instance is the right owner here. In theory anything that could reach this pointer (and can't change before marking would rescan to find this write barrier) is a valid owner. In practice, the right answer is to use the object whose visitChildren would scan this WriteBarrier. In this case the WriteBarrier is hanging off the instance so this actually the direct owner, which is always correct.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.h:
```diff
@@ -0,0 +1,311 @@
+    ImplementationPtr implementation() const { return m_implementation; }
+
+    // A JS function used in place of the builtin if it's re-exported by the module.
+    JSObject* jsRepresentative(JSGlobalObject*) const;
```
Nit: Would call this jsWrapper. That's a bit more in line with how we've named things before.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
```diff
@@ -506,7 +556,14 @@ void WebAssemblyModuleRecord::initializeExports(JSGlobalObject* globalObject)
+                // No function import means the import is a wasm builtin.
+                // The boxed callee in callLinkInfo is a WasmBuiltinCallee with a pointer to the builtin.
+                auto* callLinkInfo = m_instance->importFunctionInfo(functionIndexSpace);
+                auto* callee = std::bit_cast<Wasm::WasmBuiltinCallee*>(callLinkInfo->boxedCallee.asNativeCallee());
```
In theory we should implement a `TypeCastTraits` via `SPECIALIZE_TYPE_TRAITS_BEGIN` for wasm callee but for now we should just add an `ASSERT(callee->compilationMode() == CompilationMode::WasmBuiltinMode)`

kmiller68 commented on Source/JavaScriptCore/wasm/WasmModuleInformation.h:
```diff
@@ -171,8 +175,13 @@ struct ModuleInformation final : public ThreadSafeRefCounted<ModuleInformation>
+
     // FIXME: These should probably be FixedVectors.
     Vector<Import> imports;
+    Vector<bool> importShouldBeHidden; // filter imports[i] from the result of Module.imports(moduleObject)
```
This should be a `FixedBitVector`. ModuleInformation's can have on the order of 10,000-100,000 functions (although maybe less imports) so saving space here could matter a lot.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
```diff
@@ -1091,6 +1091,29 @@ extern "C" UGPRPair SYSV_ABI slow_path_wasm_throw_exception(CallFrame* callFrame
 
+// Similar logic to 'slow_path_wasm_throw_exception', but the exception is already sitting
+// in the VM. We don't throw, we only unwind and go to the handler.
+extern "C" UCPURegister SYSV_ABI slow_path_wasm_dispatch_exception(CallFrame* callFrame, JSWebAssemblyInstance* instance)
```
Nit: I would call this `slow_path_wasm_unwind_exception`

kmiller68 commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:
```diff
@@ -333,6 +335,9 @@ class JSWebAssemblyInstance final : public JSNonFinalObject {
     void* m_faultPC { nullptr };
+    // Used by builtin trampolines to quickly fetch callee bits to store in the call frame.
+    // The actual callees are owned by builtins. Populated by WebAssemblyModuleRecord::initializeImports().
+    CalleeBits m_builtinCalleeBits[WASM_BUILTIN_COUNT];
```
Why not have these in some global data structure that gets initialized when a native callee gets used?

ast-hugger replied:
I'll look into that, but let's make it a followup radar. That's a larger perturbation than other changes, and I'd like to land this and the other Interop PR this week, if possible.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+    }
+}
+
+const WebAssemblyBuiltinRegistry& WebAssemblyBuiltinRegistry::instance()
```
This should be called `singleton` and the implementation needs a `std::once` guard since we compile with thread safe static initializers off.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmModuleInformation.h:
```diff
@@ -171,8 +175,13 @@ struct ModuleInformation final : public ThreadSafeRefCounted<ModuleInformation>
     size_t totalFunctionSize() const { return m_totalFunctionSize; }
 
+    void applyCompileOptions(const WebAssemblyCompileOptions&);
+    bool importedStringConstantsEquals(const String& expected) const { return m_importedStringConstants && m_importedStringConstants.value() == expected; }
```
Can we `ASSERT_IMPLIES(m_importedStringConstants, !m_importedStringConstants->isUniqued());` we're strongly relying on that

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+    } while (false)
+
+#define IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope) \
+    if (scope.exception()) [[unlikely]] \
```
Does this work? I think you need to clear the exception to please the exception check validator. Are there tests that exercise this?

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+    if (!arg.isString()) [[unlikely]]
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+    auto* string = arg.toString(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
```
I think we should just rethrow the OOM here. So `RETURN_IF_EXCEPTION(scope, 0);` I would find getting an illegal argument error randomly super hard to diagnose when it was actually an OOM. In general, we're always free to OOM.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+    auto* string = arg.toString(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
+    auto view = string->view(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+    if (!arg.isString()) [[unlikely]]
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+    auto* string = arg.toString(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+    auto* string = arg.toString(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
+    auto view = string->view(globalObject);
+    IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT(scope);
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+ * is thrown. However, it does not specify what happens if the index is negative. Step 4 is not
+ * applicable because it states "Return CharCodeAt(string, index)", and CharCodeAt behaves as
+ * String.prototype.charCodeAt, which would return NaN given a negative index. I'm assuming the
+ * intended but underspecified behavior is to throw.
```
We should file a bug against the spec https://github.com/WebAssembly/js-string-builtins/issues and link to it here. What do FF/Chrome do?

ast-hugger replied:
This is pretty interesting. Chrome throws on negative values as a feature of the implementation. The int32 index is cast into uint32, so negative values are reinterpreted as positive in the range `2^31..2^32-1`. Chrome's max string length even on 64-bit systems is just short of `2^29`, so all values in that range cause an OOB failure. 

This must be why the spec is pretty sloppy here. Our max string length doesn't allow this trick, but it's correct to throw to match that behavior.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+ * The spec does not specify the interpretation of negative indices, and redundantly
+ * specifies the behavior for 'start > end' and 'start > length', which already
+ * follows from the spec for String.prototype.substring. What's implemented here
+ * is what String.prototype.substring does.
```
Ditto on filing a bug.

ast-hugger replied:
This appears to be the same phenomenon as with codePointAt, where negative index values become extremely large positive values which are OOB for all possible JS strings.

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+    JSString* leftString = left.toStringOrNull(globalObject);
+    JSString* rightString = right.toStringOrNull(globalObject);
+    auto leftView = leftString->view(globalObject);
+    auto rightView = rightString->view(globalObject);
```
These can OOM so you'll need a `RETURN_IF_EXCEPTION`

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+
+    JSString* leftString = left.toStringOrNull(globalObject);
+    JSString* rightString = right.toStringOrNull(globalObject);
```
Should be `jsCast<JSString*>(left)` or `asString(left)`

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+        THROW_ILLEGAL_ARGUMENT_EXCEPTION;
+    unsigned length = asString(arg)->length();
+    if (length > static_cast<uint64_t>(std::numeric_limits<int32_t>::max())) [[unlikely]]
+        length = std::numeric_limits<int32_t>::max();
```
This is always a no-op because `String::MaxLength == std::numeric_limits<int32_t>::max()`

kmiller68 commented on Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp:
```diff
@@ -0,0 +1,833 @@
+    auto stringValue = string->value(globalObject);
+
+    if (stringValue.data.is8Bit()) {
+        for (size_t i = 0; i < stringLength; ++i)
+            dest[i] = stringValue.data[i];
```
Is this correct for non-ASCII Latin1 characters?

ast-hugger replied:
It is, and I added a specific test just to be sure.

kmiller68 reviewed (commented):
Overall looks pretty good

kmiller68 reviewed (approved):
r=me

ast-hugger committed:
Implement Wasm js-string builtins
https://bugs.webkit.org/show_bug.cgi?id=292630
rdar://150798623

Reviewed by Keith Miller.

This patch implements the Wasm js-string builtin proposal
(https://github.com/WebAssembly/js-string-builtins).

Availability of this feature is controlled by the `useWasmJSStringBuiltins` option,
`true` by default.

The relevant WPT test expectation files are changed to expect the tests to pass.

`FixedBitVector::test()` function is marked as `const` by this patch. It does
not require a mutable `this`, and the patch needs to use it on a vector
coming from a `const ModuleInformation&`.

Canonical link: https://commits.webkit.org/298653@main
